mapiproxy/servers/default/asyncemsmdb/dcesrv_asyncemsmdb.c

```diff
@@ -0,0 +1,1230 @@
+			j += ASYNCEMSMDB_SOGO_UNDERSCORE_LEN;
+			break;
+		case ASYNCEMSMDB_AT:
+			memcpy(&_uri[j], ASYNCEMSMDB_SOGO_AT, ASYNCEMSMDB_SOGO_AT_LEN);
+			j += ASYNCEMSMDB_SOGO_AT_LEN;
+			break;
+		default:
+			_uri[j] = url_tmp[i];
+			j += 1;
+			break;
+		}
+	}
+	talloc_free(url_tmp);
+	*uri = talloc_strdup(_mem_ctx, _uri);
+	MAPISTORE_RETVAL_IF(!*uri, MAPISTORE_ERR_NO_MEMORY, NULL);
+	talloc_free(_uri);
+
+	talloc_free(mem_ctx);
+	return MAPISTORE_SUCCESS;
+}
+
+/**
+   \details Retrieve properties from mapisore object
+
+   \param mem_ctx pointer to the memory context to use for returned data memory allocation
+   \param p pointer to the private asyncemsmdb data
+   \param folderId the ID of the folder from which data are to be retrieved
+   \param properties pointer to the array of MAPI properties to retrieve
+   \param data_pointers pointer on pointer to the data to return
+   \param retvals pointers to the return valud of each data_pointers entry to return
+
+   \return MAPISTORE_SUCCESS on success, otherwise MAPISTORE error
+ */
+static enum mapistore_error get_properties_mapistore(TALLOC_CTX *mem_ctx, struct asyncemsmdb_private_data *p,
+						     uint64_t folderId, struct SPropTagArray *properties,
+						     void **data_pointers, enum MAPISTATUS *retvals)
+{
+	uint32_t			contextID;
+	enum mapistore_error		retval;
+	void				*context_object;
+	char				*uri = NULL;
+	bool				soft_deleted;
+	struct mapistore_property_data	*prop_data;
+	int				i;
+
+	/* Sanity checks */
+	MAPISTORE_RETVAL_IF(!p, MAPISTORE_ERR_NOT_INITIALIZED, NULL);
+	MAPISTORE_RETVAL_IF(!properties, MAPISTORE_ERR_INVALID_PARAMETER, NULL);
+	MAPISTORE_RETVAL_IF(!data_pointers, MAPISTORE_ERR_INVALID_PARAMETER, NULL);
+	MAPISTORE_RETVAL_IF(!retvals, MAPISTORE_ERR_INVALID_PARAMETER, NULL);
+
+	/* Retrieve the mapistore folder URI */
+	retval = mapistore_indexing_record_get_uri(p->mstore_ctx, p->username, mem_ctx, folderId, &uri, &soft_deleted);
+	MAPISTORE_RETVAL_IF(retval, retval, NULL);
+
+	/* Set or add context */
+	retval = mapistore_search_context_by_uri(p->mstore_ctx, uri, &contextID, &context_object);
+	if (retval == MAPISTORE_SUCCESS) {
+		retval = mapistore_add_context_ref_count(p->mstore_ctx, contextID);
+	} else {
+		retval = mapistore_add_context(p->mstore_ctx, p->username, uri, folderId, &contextID, &context_object);
+	}
+	MAPISTORE_RETVAL_IF(retval, retval, NULL);
+
+	prop_data = talloc_array(NULL, struct mapistore_property_data, properties->cValues);
+	MAPISTORE_RETVAL_IF(!prop_data, MAPISTORE_ERR_NO_MEMORY, NULL);
+	memset(prop_data, 0, sizeof(struct mapistore_property_data) * properties->cValues);
+
+	retval = mapistore_properties_get_properties(p->mstore_ctx, contextID, context_object, mem_ctx, properties->cValues,
+						     properties->aulPropTag, prop_data);
+	MAPISTORE_RETVAL_IF(retval, retval, prop_data);
+
+	for (i = 0; i < properties->cValues; i++) {
+		if (prop_data[i].error) {
+				if (properties->aulPropTag[i] == PidTagFolderFlags) {
```

Comment:
Could you document this special case? I don't know the reason of this condition.